src/couch_replicator_connection.erl

```diff
@@ -0,0 +1,183 @@
+        exit({ok, Pid})
+    end).
+
+handle_call({acquire, URL}, From, State) ->
+    {Pid, _Ref} = From,
+    monitor(process, Pid),
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    AsyncPid = create_worker_async(URL),
+                    true = ets:insert_new(
+                        ?MODULE,
+                        #connection{host=Host, port=Port, owner=Pid, worker=AsyncPid}
+                    ),
+                    Waiting = dict:store(AsyncPid, From, State#state.waiting),
+                    {noreply, State#state{waiting=Waiting}};
+                {[Worker], _Cont} ->
+                    ets:insert(?MODULE, Worker#connection{owner=Pid}),
+                    {reply, {ok, Worker#connection.worker}, State}
+            end
+    end.
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    case ets:lookup(?MODULE, WorkerPid) of
+        [Worker] ->
+            ets:insert(?MODULE, Worker#connection{owner=undefined});
+        [] ->
+            ok
+    end,
+    {noreply, State}.
+
+
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
+    lists:foreach(fun(Worker) ->
+        ets:insert(?MODULE, Worker#connection{owner=undefined})
+    end, Workers),
+    {noreply, State};
+
+% worker crashed
+handle_info({'EXIT', Pid, Reason}, State) ->
+    case dict:find(Pid, State#state.waiting) of
+        {ok, From} ->
+            Response = case Reason of
+                {ok, WorkerPid} ->
+                    [Worker] = ets:lookup(?MODULE, Pid),
+                    ets:insert(?MODULE, Worker#connection{worker=WorkerPid}),
+                    {ok, WorkerPid};
+                Other ->
+                    couch_log:error("Async replication connection creator died: ~p", [Other]),
+                    error
+            end,
+            gen_server:reply(From, Response),
+            {noreply, State#state{waiting=dict:erase(Pid, State#state.waiting)}};
+        error ->
+            case ets:lookup(?MODULE, Pid) of
+                [] ->
+                    ok;
+                [Worker] ->
+                    #connection{host=Host, port=Port} = Worker,
+                    couch_log:info(
+                        "Replication connection to: ~p:~p died with reason ~p",
+                        [Host, Port, Reason]
+                    ),
+                    ets:delete(?MODULE, Pid)
+            end,
+            {noreply, State}
+    end;
+
+handle_info(close_idle_connections, State) ->
+    #state{
+        close_interval=Interval,
+        timer=Timer
+    } = State,
+    Conns = ets:match_object(?MODULE, #connection{owner=undefined, _='_'}),
+    lists:foreach(fun(Conn) ->
+        ets:delete(?MODULE, Conn#connection.worker),
```

Comment:
We close all unowned connections every minute. But we don't track how long each connection has been unowned? Just checking if that is what we want. 1 minute is also the default scheduler interval. I think there could be case where we'd close connection which just got unowned, but would have gotten picked by the new group of rotated processes in the scheduler. But maybe that's too complicated. 

We could even just increase the default to be slightly longer than the default scheduling interval. So processes which would rotate back, would get a better chance to pick up their connections again.